### PR TITLE
Ben sepanski/include interface summaries [VAN-195]

### PR DESCRIPTION
**Pull request by benSepanski**
Stacked onto https://github.com/Veridise/Vanguard/pull/77 to keep PRs small.

Includes interface summaries during the driver script, and makes appropriate fixes for the Reentrancy Detector.

### CODE REVIEW

File: vanguard_driver.py [hunk 1]
```diff
@@ -240,6 +240,13 @@ def generate_solidity_llvm(args: VanguardDriverOpts) -> Path:
         cdata["name"]: cdata for cdata in summary_dict["contracts"]
     }
 
+    # Prepare summary for interfaces
+    def is_interface(summary: dict) -> bool:
+        return summary.get("kind") == "interface"
+
+    interface_summaries: List[dict] = [summary for summary in summaries.values() if is_interface(summary)]
+    interfaces_included = len(interface_summaries) == 0
```
Comment (Technius) [inline vanguard_driver.py:248]: Why is this needed? Why not always include the interfaces?

File: vanguard_driver.py [hunk 2]
```diff
@@ -255,15 +262,25 @@ def generate_solidity_llvm(args: VanguardDriverOpts) -> Path:
         # Copy .ll file to make it easier to work with
         shutil.copy(ll_file, ll_in_path, follow_symlinks=True)
 
+        # pretty-print if debugging
+        indent = None
+        if logging.getLogger().isEnabledFor(logging.DEBUG):
```
Comment (Technius) [inline vanguard_driver.py:267]: Might be better to check against `args.log_level` instead of inspecting the global root logger.

File: vanguard_driver.py [hunk 3]
```diff
@@ -255,15 +262,25 @@ def generate_solidity_llvm(args: VanguardDriverOpts) -> Path:
...
+            indent = "  "
+
         # Copy summary data from summary json
+        contracts = [summaries[contract_name]]
+        if not interfaces_included:
+            contracts.extend(interface_summaries)
```
Comment (Technius) [inline vanguard_driver.py:273]: This only appears to affect the first contract that is found.

File: vanguard_driver.py (hunk 1 above)
Comment (benSepanski) [reply vanguard_driver.py:248]: It's not necessary, it just sets `interfaces_included` to true when there are no interface summaries to include. I can change it to `False` for code clarity

File: vanguard_driver.py (hunk 3 above)
Comment (benSepanski) [reply vanguard_driver.py:273]: Yes, Vanguard has no checks for duplicate contract IDs. If I include the interface into multiple contracts then duplicate Contract objects are created in the universe

File: vanguard_driver.py (hunk 2 above)
Comment (benSepanski) [reply vanguard_driver.py]: Addressed in https://github.com/Veridise/Vanguard/pull/78/commits/088b4aba96c31da663402f1d97c631f863d16dda (used `args.logger` here since the referenced `args` is a `VanguardDriverOpts`)

File: vanguard_driver.py (hunk 3 above)
Comment (Technius) [reply vanguard_driver.py:273]: Seems like we should fix this issue in the detector program and not the driver, then. Let's fix it in a follow-up PR so we don't block this PR.

Comment (benSepanski) [reply vanguard_driver.py:273]: SGTM

File: lib/detectors/CallGraphReentrancy/ReentrancyReportFormatter.h [hunk 4]
```diff
@@ -153,8 +153,11 @@ namespace vanguard {
 
         ReentrancyReportFormatter& function(const typename Domain::Function *func) {
             assert(func != nullptr);
-            this->recordLocation(func->name());
-            return *this << func->name();
+            std::string name = func->name();
+            if(func->contract() != nullptr) {
+                name = func->contract()->name() + "." + name;
+            }
+            this->recordLocation(name); return *this << name;
```
Comment (Technius) [inline lib/detectors/CallGraphReentrancy/ReentrancyReportFormatter.h:160]: nit: need a newline here

**Review by Technius — APPROVED** [review]
LGTM.

File: lib/detectors/CallGraphReentrancy/ReentrancyReportFormatter.h (hunk 4 above)
Comment (benSepanski) [reply lib/detectors/CallGraphReentrancy/ReentrancyReportFormatter.h]: Fixed in https://github.com/Veridise/Vanguard/pull/78/commits/80b942229df799ae12de524841e3b78fd97e9abb

File: lib/analysis/interproc-analysis/call-graph/SVFCallGraphBuilder.h [hunk 5]
```diff
@@ -104,14 +121,22 @@ namespace vanguard {
             }
         };
 
+        static std::shared_ptr<SVF::PTACallGraphNode> externalNode;
```
[the remark this answers is not in the document]
Comment (benSepanski) [reply lib/analysis/interproc-analysis/call-graph/SVFCallGraphBuilder.h:124]: Currently, all methods of SVFCallGraphBuilder are static, so the class is never actually instantiated.
